plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,221 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
+	}
+	datetime, err := time.ParseInLocation("060102 15:04:05", fmt.Sprintf("%s %s", fields[0][1:], fields[1]), time.Local)
+	if err != nil {
+		return 0, false, err
+	}
+	return datetime.Unix(), true, nil
+}
+
+func parseBinlogEventStartPosInLine(line string) (startPos int64, found bool, err error) {
+	// The mysqlbinlog output will contains a line starting with "# at 35065", which is the binlog event's start position.
+	if !strings.HasPrefix(line, "# at ") {
+		return 0, false, nil
+	}
+	// This is the line containing the start position of the binlog event.
+	fields := strings.Fields(line)
+	if len(fields) != 3 {
+		return 0, false, fmt.Errorf("unexpected mysqlbinlog output line %q when parsing binlog event start position", line)
+	}
+	startPos, err = strconv.ParseInt(fields[2], 10, 0)
+	if err != nil {
+		return 0, false, err
+	}
+	return startPos, true, nil
+}
+
+// Parse the first binlog eventTs of a local binlog file.
+func (r *Restore) parseFirstLocalBinlogEventTs(ctx context.Context, fileName string) (eventTs int64, err error) {
+	args := []string{
+		// Local binlog file path.
+		path.Join(r.binlogDir, fileName),
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	// s.Scan() could panic, so we should recover and handle this.
+	defer func() {
+		if r := recover(); r != nil {
+			var ok bool
+			err, ok = r.(error)
+			if !ok {
+				err = fmt.Errorf("%v", r)
+			}
+			log.Error("bufio.Scanner panic", zap.Error(err))
+		}
+	}()
+	var found bool
+	for s.Scan() {
+		line := s.Text()
+		eventTs, found, err = parseBinlogEventTsInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog eventTs from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		break
+	}
+
+	return eventTs, nil
+}
+
+// Use command like mysqlbinlog --start-datetime=targetTs+1 binlog.000001 to parse the first binlog event ts after targetTs.
+// The --start-datetime is inclusive, so we use targetTs+1 as the argument, which gives the position after all events before targetTs (inclusive) are replayed.
+// TODO(dragonly): Optimize to kill mysqlbinlog process after we parsed the first binlog event ts after targetTs.
+// The current implementation will parse the entire binlog file. This optimization may significantly reduce CPU/memory/disk IO usage.
+// TODO(dragonly): Add integration test.
+func (r *Restore) getFirstBinlogEventStartPosAfterTs(ctx context.Context, binlogFile BinlogFile, targetTs int64) (startPos int64, err error) {
```

Comment:
getBinlogEventPositionAtOrAfterTs()

Reply:
I prefer `getBinlogEventPositionAfterTs()`. The reason is the same as I use `--start-datetime targetTs+1`: I want to replay binlog to targetTs (inclusive).